kore/src/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,440 @@
+termToExistentials
+    :: TermLike RewritingVariableName
+    -> (TermLike RewritingVariableName, [ElementVariable RewritingVariableName])
+termToExistentials (TermLike.Exists_ _ v term) =
+    fmap (v :) (termToExistentials term)
+termToExistentials term = (term, [])
+
+forgetSimplified :: Implication modality -> Implication modality
+forgetSimplified implication' =
+    implication'
+        { left = Pattern.forgetSimplified left
+        , right = OrPattern.forgetSimplified right
+        }
+  where
+    Implication { left, right } = implication'
+
+{- | Ensure that the 'Implication''s bound variables are fresh.
+
+The 'existentials' should not appear free on the left-hand side so that we can
+freely unwrap the right-hand side as needed.
+
+See also: 'refreshExistentials'
+
+ -}
+assertRefreshed :: HasCallStack => Implication modality -> a -> a
+assertRefreshed implication'@Implication{ existentials } =
+    assert (isFreeOf implication' (Set.fromList $ inject <$> existentials))
+
+{- | Refresh the 'existentials' of the 'Implication'.
+ -}
+refreshExistentials :: Implication modality -> Implication modality
+refreshExistentials = snd . refreshRule mempty
+
+instance UnifyingRule (Implication modality) where
+    type UnifyingRuleVariable (Implication modality) = RewritingVariableName
+
+    matchingPattern implication' =
+        Pattern.term left
+      where
+        Implication { left } = implication'
+
+    precondition implication' =
+        Condition.toPredicate . Pattern.withoutTerm $ left
+      where
+        Implication { left } = implication'
+
+    refreshRule stale implication' =
+        do
+            let variables = freeVariables implication' & FreeVariables.toSet
+            renaming <- refreshVariables' variables
+            let existentials' = Set.fromList (inject <$> existentials)
+            renamingExists <- refreshVariables' existentials'
+            let subst = TermLike.mkVar <$> renaming
+                refreshedImplication =
+                    implication'
+                    & renameExistentials renamingExists
+                    & substitute subst
+            -- Only return the renaming of free variables.
+            -- Renaming the bound variables is invisible from outside.
+            pure (renaming, refreshedImplication)
+        & flip evalState (FreeVariables.toNames stale)
+      where
+        refreshVariables' variables = do
+            staleNames <- State.get
+            let renaming = refreshVariables staleNames variables
+                staleNames' = Set.map variableName variables
+                staleNames'' =
+                    Map.elems renaming
+                    & foldMap FreeVariables.freeVariable
+                    & FreeVariables.toNames
+            State.put (staleNames <> staleNames' <> staleNames'')
+            pure renaming
+        Implication { existentials } = implication'
+
+mkGoal :: Implication modality  -> Implication modality 
+mkGoal implication' =
```

Comment:
A better name for this would be `resetConfigVariables` because it is essentially just calling `resetConfigVariable` on every component of the `Implication`.